netty/src/main/java/io/grpc/netty/Utils.java

```diff
@@ -65,11 +72,50 @@
   public static final AsciiString TE_TRAILERS = AsciiString.of(GrpcUtil.TE_TRAILERS);
   public static final AsciiString USER_AGENT = AsciiString.of(GrpcUtil.USER_AGENT_KEY.name());
 
-  public static final Resource<EventLoopGroup> DEFAULT_BOSS_EVENT_LOOP_GROUP =
-      new DefaultEventLoopGroupResource(1, "grpc-default-boss-ELG");
+  public static final Resource<EventLoopGroup> NIO_BOSS_EVENT_LOOP_GROUP
+      = new DefaultEventLoopGroupResource(1, "grpc-nio-boss-ELG", NioEventLoopGroup.class);
+  public static final Resource<EventLoopGroup> NIO_WORKER_EVENT_LOOP_GROUP
+      = new DefaultEventLoopGroupResource(0, "grpc-nio-worker-ELG", NioEventLoopGroup.class);
+  public static final Resource<EventLoopGroup> DEFAULT_BOSS_EVENT_LOOP_GROUP;
+  public static final Resource<EventLoopGroup> DEFAULT_WORKER_EVENT_LOOP_GROUP;
 
-  public static final Resource<EventLoopGroup> DEFAULT_WORKER_EVENT_LOOP_GROUP =
-      new DefaultEventLoopGroupResource(0, "grpc-default-worker-ELG");
+  private static final Class<? extends ServerChannel> defaultServerChannelType;
+  private static final Class<? extends Channel> defaultClientChannelType;
+
+  static {
+    // Decide default channel types and EventLoopGroup based on Epoll availability
+    Class<? extends ServerChannel> serverChannelType = NioServerSocketChannel.class;
+    Class<? extends Channel> clientChannelType = NioSocketChannel.class;
+    Class<? extends EventLoopGroup> eventLoopGroupType = NioEventLoopGroup.class;
+    boolean epollSuccessful = false;
+
+    try {
+      if (isEpollAvailable()) {
+        serverChannelType = epollServerChannelType();
+        clientChannelType = epollChannelType();
+        // load EventLoopGroup to make sure all defaults are using same type.
+        eventLoopGroupType = epollEventLoopGroupType();
+        epollSuccessful = true;
+      }
+    } catch (Exception e) {
```

Comment:
Log this. It looks like it should be a WARNING, since the code above should not throw.

Although to make the code more obvious, we could really just let the exceptions propagate (again, since it shouldn't throw). That would allow you to assign directly to the `final` fields, as if something bad happens you don't need to worry about assigning the rest or getting into a weird state (because the program will crash).